File: minFallingPathSum.js
// Time Complexity : O(n^2) where n is length of matrix
// Space Complexity : O(n^2) where n is length of matrix
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No

/**
 * @param {number[][]} matrix
 * @return {number}
 */
var minFallingPathSum = function(matrix) {
    let n = matrix.length;
    let dp = [new Array(n)];
    dp[0] = matrix[0];
    for(let i=1; i<n; i++) {
        for(let j=0; j<n; j++) {
            if(!dp[i]) {
                dp[i] = new Array(n);
            }
            dp[i][j] = dp[i-1][j];
            if(j - 1 >= 0) dp[i][j] = Math.min(dp[i][j], dp[i-1][j-1]);
            if(j + 1 < n) dp[i][j] = Math.min(dp[i][j], dp[i-1][j+1]);
            dp[i][j] += matrix[i][j]; 
        }
    }
    
    let res = Infinity;
    for (let j = 0; j < n; j++) {
        res = Math.min(res, dp[n-1][j]);
    }
    return res;
};

File: minFallingPathSum.go
// Time Complexity : O(n^2)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None

/*

Approach:
We use a bottom up DP approach where we consider that we're going from the
"leaf" of the decision tree of which path to take, and work our way up towards
the first set of nodes of that tree (which is the first row of our matrix). We
keep accumulating the minimum sum at every row considering the path traversal
constraints (i.e., which cell to choose) imposed by the problem statement.

*/

package main

import "math"

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func minFallingPathSum(matrix [][]int) int {
	if len(matrix) == 0 {
		return 0
	}

	n := len(matrix)

	for i := n - 2; i >= 0; i-- { // start from last but one row
		for j := 0; j <= n-1; j++ { // do all columns
			if j == 0 { // first column
				matrix[i][j] += min(matrix[i+1][j], matrix[i+1][j+1])
			} else if j == n-1 { // last column
				matrix[i][j] += min(matrix[i+1][j], matrix[i+1][j-1])
			} else { // every other cell
				matrix[i][j] += min(
					matrix[i+1][j-1],
					min(matrix[i+1][j], matrix[i+1][j+1]))
			}
		}
	}

	res := math.MaxInt
	for _, e := range matrix[0] {
		res = min(res, e)
	}

	return res
}
